src/btree/bt_walk.c

```diff
@@ -132,6 +131,43 @@ __page_ascend(WT_SESSION_IMPL *session,
 }
 
 /*
+ * __page_descend --
+ *	Descend the tree one level.
+ */
+static void
+__page_descend(WT_SESSION_IMPL *session,
+    WT_REF *ref, WT_PAGE_INDEX **pindexp, uint32_t *slotp, bool prev)
+{
+	WT_PAGE_INDEX *pindex;
+
+	/*
+	 * Ref is a child page into which we're descending, and on which we
+	 * have a hazard pointer.
+	 */
+	for (;; __wt_yield()) {
+		WT_INTL_INDEX_GET(session, ref->page, pindex);
+		*slotp = prev ? pindex->entries - 1 : 0;
+
+		/*
+		 * When internal pages split, the WT_REF structures being moved
+		 * are updated first. If the last WT_REF on the page into which
+		 * we're descending doesn't reference the page as its home, the
+		 * page has split and a "previous" cursor moving to the last
+		 * slot on the page would be in the wrong part of the namespace.
+		 * The test isn't necessary for a "next" cursor because we do
+		 * right-hand splits on internal pages and the first part of the
+		 * page's namespace won't have changed as part of the split.
+		 * Rather than testing the direction boolean, do the test the
+		 * "previous" cursor movement requires, it's not wrong for the
+		 * "next" cursor movement.
+		 */
+		if (pindex->index[*slotp]->home == ref->page)
+			break;
```

Comment:
@keithbostic, this seems related to `__wt_split_intl_race` -- could we save the `pindex` from the ascend to reuse the same test?